tianshou/env/vecenv/shmemenv.py

```diff
@@ -0,0 +1,158 @@
+        BaseVectorEnv.__init__(self, env_fns)
+        self.closed = False
+        self._setup_obs_space(env_fns[0])
+        self.obs_bufs = [
+            {k: Array(_NP_TO_CT[self.obs_dtypes[k].type], int(
+                np.prod(self.obs_shapes[k]))) for k in self.obs_keys}
+            for _ in range(self.env_num)]
+        self.parent_remote, self.child_remote = \
+            zip(*[Pipe() for _ in range(self.env_num)])
+        self.processes = [
+            Process(target=_shmem_worker, args=(
+                parent, child, CloudpickleWrapper(env_fn),
+                obs_buf, self.obs_shapes,
+                self.obs_dtypes, self.obs_keys), daemon=True)
+            for (parent, child, env_fn, obs_buf) in zip(
+                self.parent_remote, self.child_remote, env_fns, self.obs_bufs)
+        ]
+        for p in self.processes:
+            p.start()
+        for c in self.child_remote:
+            c.close()
+
+    def step(self,
+             action: np.ndarray,
+             id: Optional[Union[int, List[int]]] = None
+             ) -> Tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray]:
+        if id is None:
+            id = range(self.env_num)
+        elif np.isscalar(id):
+            id = [id]
+        assert len(action) == len(id)
+        for i, j in enumerate(id):
+            self.parent_remote[j].send(['step', action[i]])
+        result = []
+        for i in id:
+            obs, rew, done, info = self.parent_remote[i].recv()
+            obs = self._decode_obs(obs, i)
+            result.append((obs, rew, done, info))
+        obs, rew, done, info = map(np.stack, zip(*result))
+        return obs, rew, done, info
+
+    def reset(self, id: Optional[Union[int, List[int]]] = None) -> np.ndarray:
+        if id is None:
+            id = range(self.env_num)
+        elif np.isscalar(id):
+            id = [id]
+        for i in id:
+            self.parent_remote[i].send(['reset', None])
+        obs = np.stack(
+            [self._decode_obs(self.parent_remote[i].recv(), i) for i in id])
+        return obs
+
+    def _setup_obs_space(self, env):
+        dummy = env()
+        obs_space = dummy.observation_space
+        dummy.close()
+        del dummy
+        if isinstance(obs_space, gym.spaces.Dict):
+            assert isinstance(obs_space.spaces, OrderedDict)
+            subspaces = obs_space.spaces
+        elif isinstance(obs_space, gym.spaces.Tuple):
+            assert isinstance(obs_space.spaces, tuple)
+            subspaces = {i: obs_space.spaces[i]
+                         for i in range(len(obs_space.spaces))}
+        else:
+            subspaces = {None: obs_space}
+        self.obs_keys = []
+        self.obs_shapes = {}
+        self.obs_dtypes = {}
+        for key, box in subspaces.items():
+            self.obs_keys.append(key)
+            self.obs_shapes[key] = box.shape
+            self.obs_dtypes[key] = box.dtype
```

Comment:
I think a better way is to use ``.observation_space.sample()`` to get a sample of observation. And it seems current implementation does not deal with recurse dict. Below is a code snippet you can use (decode looks the same). Then your implementation is different from open ai baseline :) 

```python
obs = env.observation_space.sample()
def encode(obs):
    if isinstance(obs, np.ndarray):
        shm = Array(xxx)
    if isinstance(obs, tuple):
        shm = (encode(i) for i in obs)
    if isinstance(obs, dict):
        shm = {k: encode(v) for k, v in obs.items()}
    return shm
shm = encode(obs)
```

Reply:
Thanks a lot. Yes recurse dict is not supported, but so far I haven't seen any case in which the observation is a recurse dict. What's more, it seems that standard gym env only supports the observation to be dict of numpy  arrayor a single numpy array. So I think maybe it is unnecessary to change this part just to be different with openai.
sample() looks good.

Reply:
I think it should be done

Reply:
Yeah, okay. No harm anyway. Working on it. Thanks for suggestions.

Reply:
> but so far I haven't seen any case in which the observation is a recurse dict

Here is a case of recurse dict from the [MineRL](https://minerl.io/docs/index.html) environment: 
```python
Dict(inventory:Dict(coal:Box(low=0, high=2304, shape=()), cobblestone:Box(low=0, high=2304, shape=())), pov:Box(low=0, high=255, shape=(64, 64, 3)))
```

Reply:
> Thanks a lot. Yes recurse dict is not supported, but so far I haven't seen any case in which the observation is a recurse dict.

In my project, mine are too. And it is pretty common. I just you have only seen some toy examples.